repository-meta-analyzer/src/main/java/org/dependencytrack/repometaanalyzer/repositories/DepsDevMetaAnalyzer.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * This file is part of Dependency-Track.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) OWASP Foundation. All Rights Reserved.
+ */
+package org.dependencytrack.repometaanalyzer.repositories;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.Optional;
+
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.dependencytrack.persistence.model.Component;
+import org.dependencytrack.persistence.model.RepositoryType;
+import org.dependencytrack.repometaanalyzer.model.MetaModel;
+import org.dependencytrack.repometaanalyzer.util.PurlUtil;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.packageurl.PackageURL;
+
+public class DepsDevMetaAnalyzer extends AbstractMetaAnalyzer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DepsDevMetaAnalyzer.class);
+    private static final String DEFAULT_BASE_URL = "https://api.deps.dev/v3alpha";
+    private static final String API_URL = "/purl/%s";
+    private static final String SOURCE_REPO = "SOURCE_REPO";
+
+    DepsDevMetaAnalyzer() {
+        this.baseUrl = DEFAULT_BASE_URL;
+    }
+
+    @Override
+    public boolean isApplicable(Component component) {
+        return component.getPurl() != null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public RepositoryType supportedRepositoryType() {
+        return null; // Supported values for type are cargo, golang, maven, npm, nuget and pypi. 
+    }
```

Comment:
The analyzer would still need to be registered with `RepositoryAnalyzerFactory`, based on the PURL types it supports:

https://github.com/DependencyTrack/hyades/blob/941dd0ffa8bc05d3b8d6340bc2c7d5050e56c025/repository-meta-analyzer/src/main/java/org/dependencytrack/repometaanalyzer/repositories/RepositoryAnalyzerFactory.java#L33-L44

The current model assumes at most one analyzer per PURL type though, so we'd need to adjust this to support multiple. Question then is, should one take priority over the other? Do we execute all applicable analyzers, and if so, how do we merge all results back into one?

Reply:
I'm wondering if perhaps we should switch to deps.dev entirely for all public components. The `internal` status is already provided to the repository meta analyzer as per Protobuf definition: https://github.com/DependencyTrack/hyades/blob/941dd0ffa8bc05d3b8d6340bc2c7d5050e56c025/proto/src/main/proto/org/dependencytrack/repometaanalysis/v1/repo_meta_analysis.proto#L45-L47

In that case, we would not only source the repository from deps.dev, but also:

* Latest version
* Publish timestamp of latest version
* Publish timestamp of current version
* Hashes of current version (not sure if deps.dev provides that?)

I think that would be simpler than trying to support multiple analyzers per PURL.

Does that sound reasonable? Granted that change would be a bit larger.

Reply:
> In that case, we would not only source the repository from deps.dev, but also:
>
> - Latest version
> - Publish timestamp of latest version
> - Publish timestamp of current version
> - Hashes of current version (not sure if deps.dev provides that?)

DevDeps accepts purls with and without version tags but returns different content:
If the version **is not** part of the purl we get a json object containing information about all available versions. If the version **is** part of the purl, we get the information about the given version enriched with additional data such as SOURCE_REPO. From this we could extract

- Given version
- Publish timestamp of this version
- version's source repo (if provided)

In DT we have purls with version tags. Finding out the **latest** version of a package would be complicated given the way DevDeps purlLookup works. We would have to query DevDeps twice, once without version tag and a second time with version tag. Probably not a good idea...